src/listen/ldap_sync/active_directory.c

```diff
@@ -0,0 +1,684 @@
+/*
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ * @file active_directory.c
+ * @brief LDAP sync callback functions for Active Directory servers.
+ *
+ * @copyright 2022 Network RADIUS SARL (legal@networkradius.com)
+ */
+
+#define LOG_PREFIX "ldap_sync_ad"
+
+#include "active_directory.h"
+#include <freeradius-devel/util/debug.h>
+
+/** Check that a particular attribute is included in the query attribute list
+ *
+ * @param[in] attrs	list to check
+ * @param[in] attr	to look for
+ * @param[in] msg	to include in the warning
+ */
+static void ldap_query_attrs_check(char const **attrs, char const *attr, char const *msg)
+{
+	size_t		len, i;
+
+	if (!attr) return;
+
+	len = talloc_array_length(attrs);
+
+	/*
+	 *	No attributes specified - so all will be returned
+	 */
+	if (len == 0) return;
+
+	for (i = 0; i < len; i++) {
+		if (!attrs[i]) continue;
+		if (strcasecmp(attrs[i], attr) == 0) return;
+		if (strcasecmp(attrs[i], "*") == 0) return;
+	}
+
+	WARN("Attribute %s %s, but not in requested attributes", attr, msg);
+}
+
+/** Terminal characters for parsing LDAP filters
+*/
+static const fr_sbuff_term_t filter_terms = FR_SBUFF_TERMS(
+	L("="),
+	L("<"),
+	L(">"),
+	L("~"),
+	L(":"),
+	L(")"),
+	L("("),
+	L("&"),
+	L("|"),
+	L("!")
+);
+
+/** Parse individual LDAP filter nodes
+ *
+ * @param[in,out] node	to populate with parsed filter
+ * @param[in] sbuff	pointing to filter to parse
+ * @param[in] depth	to indent debug output, indicating nesting of groups
+ *
+ * @return
+ *  - 0 on success
+ *  - < 0 on error
+ *  - 1 if the final advance of the sbuff moved forward.
+ */
+static int ldap_filter_parse_node(ldap_filter_t *node, fr_sbuff_t *sbuff, int depth, char const **attrs)
+{
+	fr_sbuff_marker_t	filter_marker, attr_s_marker, attr_e_marker, value_marker;
+	ldap_filter_t		*child_node;
+
+	/*
+	 *	Firstly, look for the characters which indicate the start of a group of filters
+	 *	to be combined with a logical operator.
+	 */
+	switch(fr_sbuff_current(sbuff)[0]) {
```

Comment:
Just deref with `*`, `[0]` encourages people reading the code to access offsets of current which is unsafe.